File: src/algorithms/recursion/permutations/solution2.go
package main

// O(n*n!) time | O(n*n!) space
func GetPermutations(array []int) [][]int {
	permutations := [][]int{}
	permutationsHelper(0, array, &permutations)
	return permutations
}

func permutationsHelper(i int, array []int, permutations *[][]int) {
	if i == len(array)-1 {
		newPerm := make([]int, len(array))
		copy(newPerm, array)
		*permutations = append(*permutations, newPerm)
		return
	}
	for j := i; j < len(array); j++ {
		swap(array, i, j)
		permutationsHelper(i+1, array, permutations)
		swap(array, i, j)
	}
}

func swap(array []int, i, j int) {
	array[i], array[j] = array[j], array[i]
}

File: src/algorithms/recursion/permutations/solution2.ts
// O(n*n!) time | O(n*n!) space
export function getPermutations(array: number[]) {
  const permutations: number[][] = [];
  permutationsHelper(0, array, permutations);
  return permutations;
};

function permutationsHelper(
  i: number,
  array: number[],
  permutations: number[][]
) {
	if (i === array.length - 1) {
		permutations.push(array.slice());
	} else {
		for (let j = i; j < array.length; j++) {
			swap(i, j, array);
			permutationsHelper(i + 1, array, permutations);
			swap(i, j, array);
		};
	};
};

function swap(
  i: number,
  j: number,
  array: number[]
) {
	const temp = array[i];
	array[i] = array[j];
	array[j] = temp;
};
